File: LeetCode/Study Plan/Practice Problems/Medium/Construct A Quad Tree/Java/constructQuadTree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
};
*/

class Solution {
    public Node construct(int[][] grid) {
        int size = grid.length;
        return recur(0, 0, grid, size);
    }

    public Node recur(int x, int y, int[][] grid, int size) {
        if (isSameValue(x, y, grid, size)) {
            return new Node(grid[x][y]==1, true);
        }else {
            Node root = new Node(false, false);
            root.topLeft = recur(x, y, grid, size/2);
            root.topRight = recur(x, y+size/2, grid, size/2);
            root.bottomLeft = recur(x+size/2, y, grid, size/2);
            root.bottomRight = recur(x+size/2, y+size/2, grid, size/2);
            return root;
        }
    }

    public boolean isSameValue(int x, int y, int[][] grid, int size) {
        for (int i=x; i<x+size; i++) {
            for(int j=y; j<y+size; j++) {
                if (grid[x][y] != grid[i][j]) return false;
            }
        }

        return true;
    }
}

File: LeetCode/Study Plan/Practice Problems/Medium/Construct A Quad Tree/Go/constructQuadTree.go
package main

import "fmt"

/**
 * Definition for a QuadTree node.
 * type Node struct {
 *     Val bool
 *     IsLeaf bool
 *     TopLeft *Node
 *     TopRight *Node
 *     BottomLeft *Node
 *     BottomRight *Node
 * }
 */

type Node struct {
	Val         bool
	IsLeaf      bool
	TopLeft     *Node
	TopRight    *Node
	BottomLeft  *Node
	BottomRight *Node
}

func New(val, isLeaf bool) *Node { return &Node{val, isLeaf, nil, nil, nil, nil} }

func construct(grid [][]int) *Node {
	size := len(grid)
	return recur(0, 0, grid, size)
}

func recur(x, y int, grid [][]int, size int) *Node {
	if isSameValue(grid, x, y, size) {
		return New(grid[x][y] == 1, true)
	} else {
		//call recur over the four submatrices
		root := New(false, false)
		root.TopLeft = recur(x, y, grid, size/2)
		root.TopRight = recur(x, y+size/2, grid, size/2)
		root.BottomLeft = recur(x+size/2, y, grid, size/2)
		root.BottomRight = recur(x+size/2, y+size/2, grid, size/2)
		return root
	}
}

func isSameValue(grid [][]int, x, y, size int) bool {
	for i := x; i < x+size; i++ {
		for j := y; j < y+size; j++ {
			if grid[x][y] != grid[i][j] {
				return false
			}
		}
	}
	return true
}

func main() {
	fmt.Println(construct([][]int{
		{0, 1}, {1, 0},
	}))
}
